### PR TITLE
Query based routing.

### PR DESCRIPTION
**Pull request by sarvjeetsinghghotra**
Added a `query` function to allow query based routing. It takes a `dict(String, String)` as input for target query.

Following assumptions are made:
1. `request` query and target query should be exactly same i.e both should have same key and value pairs.
2. target query can have empty value for some key for e.g `dict("one" => "1", "two" => "")` in this case it matches for any value of `two` but `two` key must be there in `request`  with some value. (Not sure about this case so need some inputs)

### CODE REVIEW

File: src/routing.jl [hunk 1]
```diff
@@ -44,6 +45,23 @@ page(p::AbstractString, app...) = page(splitpath(p), app...)
 page(app...) = page([], app...)
 page(app::Function, p) = page(p, app)
 
+# Query routing
+
+function matchquery(q, req)
+  qdict = parsequerystring(req[:query])
+  length(q) != length(qdict) && return false
+  for pair in q
```
Comment (MikeInnes) [inline src/routing.jl]: `for (key, value) in ...` might be nicer here

**Comment by MikeInnes** [comment]
A good idea and a good pr! Just a couple comments to look at.

File: src/routing.jl [hunk 2]
```diff
@@ -19,6 +19,7 @@ function matchpath(target, path)
   params = d()
   for i = 1:length(target)
     if startswith(target[i], ":")
+      params = d()
```
Comment (MikeInnes) [inline src/routing.jl]: What's this for?

**Comment by sarvjeetsinghghotra** [comment]
Fixed the problems that you pointed out, thank you. Is there a problem if it didn't pass the Travis CI build for Julia-0.5 and 0.4 version ?
@MikeInnes

**Comment by sarvjeetsinghghotra** [comment]
@MikeInnes This will help me to prove my understanding of the Mux project for the GSoC. So could you please review the updated (fixed the problems that you pointed) PR soon? 

Also, Is it OK if it doesn't pass the Travis CI build for Julia 0.5 and 0.4 version ?

File: src/routing.jl [hunk 3]
```diff
@@ -44,6 +44,23 @@ page(p::AbstractString, app...) = page(splitpath(p), app...)
 page(app...) = page([], app...)
 page(app::Function, p) = page(p, app)
 
+# Query routing
+
+function matchquery(q, req)
+  qdict = parsequerystring(req[:query])
```
Comment (jpsamaroo) [inline src/routing.jl]: I'm trying to see if I can rebase this PR on top of master, but this one call `parsequerystring` doesn't exist, and I'm having trouble locating it in the git history. Does anyone know what this function is supposed to do? @sarvjeetsinghghotra @MikeInnes

Comment (aviks) [reply src/routing.jl]: https://github.com/JuliaWeb/HttpCommon.jl/search?q=parsequerystring&unscoped_q=parsequerystring

**Comment by jpsamaroo** [comment]
So this PR is going to require adding a dependency on HttpCommon; are we okay with that? It needs URIParser and Nullables. If so, we should just be able to rebase this on top of Mux master (not sure if that's what Avik's second commit is doing or if it's just a merge) and then add in the necessary REQUIRE lines and `using` statements, then this PR should function just fine. I'm testing that out and will report back if I run into any issues.

**Comment by hustf** [comment]
I think functionality like parsequerystring belong in an optional utilty package like HttpCommon. Moving "everything" to Http.jl is a way to facilitate making some larger structure changes which were necessary.  But this will be hard to maintain in one package over time since too few people know everything. That said,  have you checked if similar functionality now exists in Http.jl?

**Comment by aviks** [comment]
No, we cannot take dependency on HTTPCommon, that package is effectively deprecated. There should be equivalent functionality in HTTP.jl

To @hustf's point, that may be the case eventually, but for now everything in HTTP.jl is simpler to maintain.

**Comment by aviks** [comment]
https://github.com/JuliaWeb/HTTP.jl/blob/master/src/URIs.jl#L287

**Comment by aviks** [comment]
continued in #81 . Thanks @sarvjeetsinghghotra for getting this started.

**Pull request closed without merging**